### PR TITLE
docs: remediate admin CLI trifecta audit findings

### LINKED ISSUE #1007: Trifecta Audit: Admin CLI — 35 findings across 35 requirements
# Admin CLI Trifecta Audit — Requirements → Design/Validation

**Audit date:** 2026-05-19
**Documents audited:** `admin-requirements.md`, `admin-design.md`, `admin-validation.md`
**Requirements audited:** 35 | **Findings:** 35 | **Clean:** 2 (5.7%)

## Summary

Admin CLI has the weakest trifecta coverage. The systemic issue is that validation tests call the RPC/client layer, not the CLI surface — prompts, output format, error messages, and interactive behaviors are almost entirely untested at the acceptance-criteria level. Only ADMIN-0106 and ADMIN-0603 are fully clean.

### Systemic Pattern
The dominant gap is D7: validation tests exist for every requirement (no D2 gaps), but tests exercise the gRPC client wrapper rather than the CLI binary. This means CLI-surface behaviors (stderr prompts, stdout formatting, `--format json`, `--yes` bypass, non-TTY refusal) are systematically untested.

### F-ADM-001: ADMIN-0101 — D7 — High
Default endpoint, `--socket`, Windows `ERROR_PIPE_BUSY` retry, and endpoint-in-error-text ACs not covered.
- **Evidence:** Req `admin-requirements.md:72-87`; tests check connection success/failure only.

### F-ADM-002: ADMIN-0103 — D7 — High
Prompt shape (`[y/N]:`) not asserted; `y`/`Y` acceptance behavior untested.
- **Evidence:** Req `admin-requirements.md:117-129`; test checks `n` abort and `--yes` only.

### F-ADM-003: ADMIN-0104 — D7 — High
Non-TTY success with `--yes` untested.
- **Evidence:** Req `admin-requirements.md:138-146`; test checks refusal without `--yes` only.

### F-ADM-004: ADMIN-0203 — D7 — High
`node remove` prompt text and interactive `y` confirmation path under-tested.
- **Evidence:** Req `admin-requirements.md:293-299`.

### F-ADM-005: ADMIN-0204 — D7 — High
`factory-reset` CLI process tests with/without `--yes` missing.
- **Evidence:** Req `admin-requirements.md:309-315`; test checks `factory_reset("node-id")` only.

### F-ADM-006: ADMIN-0300 — D5/D7 — High
Validation says "Build a minimal CBOR program image" but requirement says "BPF ELF object file". CLI file-I/O layer and local error untested.
- **Evidence:** Req `admin-requirements.md:327-336`; validation `admin-validation.md:327-331`.

### F-ADM-007: ADMIN-0303 — D7 — High
`program remove` CLI confirmation/output tests missing.
- **Evidence:** Req `admin-requirements.md:379-386`; test checks RPC removal only.

### F-ADM-008: ADMIN-0501 — D7 — High
All three ACs (confirmation, file read, success report) missing at CLI layer for state import.
- **Evidence:** Req `admin-requirements.md:491-499`; test checks API-level import only.

### F-ADM-009: ADMIN-0502 — D7 — High
Passphrase priority order (`--passphrase` → `SONDE_PASSPHRASE` → TTY prompt) and no-TTY failure untested. Only empty-passphrase rejection covered.
- **Evidence:** Req `admin-requirements.md:509-523`; validation `admin-validation.md:467-469`.

### F-ADM-010: ADMIN-0700 — D1/D7 — High
Design coverage incomplete for core interactive pairing flow. Validation covers only subset of six ACs.
- **Evidence:** Req `admin-requirements.md:608-624`; design gives one-line trace only (`admin-design.md:239`).

### F-ADM-011: ADMIN-0701/0703 — D7 — High
Positive destructive behavior (close pairing window, revoke phone) not validated. Tests only cover error paths.
- **Evidence:** Req `admin-requirements.md:633-672`; validation checks error cases only.

## Medium-Severity Findings (18)

ADMIN-0102, 0105, 0107, 0200, 0201, 0202, 0301, 0302, 0400, 0402, 0403, 0500, 0600, 0601, 0602, 0702, 0800, 0802 — all D7, all following the same pattern: validation tests the RPC layer but not CLI output format, text/JSON parity, or edge-case rendering.

## Low-Severity Findings (6)

ADMIN-0100, 0401, 0801 (CLI output assertions missing), plus three others with minor traceability gaps.

## Remediation

The systemic fix is to add CLI-process integration tests that invoke the `sonde-admin` binary, capture stdout/stderr, and assert prompt text, output format, error messages, and interactive behavior. This could be a single test harness covering all subcommands.

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Summary

Remediates all 35 findings from the admin CLI trifecta audit (#1007) by updating
\dmin-design.md\ and \dmin-validation.md\. Requirements are unchanged.

### Design changes (\dmin-design.md\)

- **Added §6a** — \\Pairing start — interactive event loop\\ describing the
  \\OpenBlePairing\\ streaming RPC lifecycle, event variant→CLI action mapping,
  passkey confirmation flow (6-digit zero-padded, stderr prompt, \\ConfirmBlePairing\\),
  non-TTY behavior, and loop exit on \\WindowClosed\\. Closes D1 gap for ADMIN-0700.

### Validation changes (\dmin-validation.md\)

- **Augmented 5 existing tests**: T-0100, T-0103, T-0107, T-0108, T-0300
  - T-0107: Added \\[y/N]:\\ prompt shape assertion, \\y\\/\\Y\\ acceptance path
  - T-0108: Added \\--yes\\ success in non-TTY mode
  - T-0300: Fixed D5 mismatch (CBOR→ELF per requirement)
  - T-0100/T-0103: Reclassified as CLI process tests
- **Added 28 new test cases** covering CLI process-level behavior for all subcommands
- **Total test cases**: 69 (up from ~40)
- **CLI process tests**: 41 (up from ~9)

### Findings addressed

| Severity | Count | Pattern |
|----------|-------|---------|
| High | 11 | F-ADM-001 through F-ADM-011 |
| Medium | 18 | D7 (RPC-level tests only) across 18 requirements |
| Low | 6 | Minor traceability gaps |

### Systemic fix

The dominant D7 pattern — validation tests exercising the gRPC client wrapper rather
than the CLI binary — is resolved by adding CLI process tests that invoke \\sonde-admin\\
via \\std::process::Command\\ and assert on stdout, stderr, and exit codes.

### Known limitations

- Modem status/scan positive-path tests require a modem transport fixture (deferred)
- BLE pairing positive-path tests require a pairing harness (T-0703a is structural)
- Phone revocation positive-path test requires a phone registration fixture (T-0702a is structural)

Closes #1007

### CODE REVIEW

File: docs/admin-design.md [hunk 1]
```diff
@@ -246,6 +246,46 @@ hash fields returned by the node-oriented RPC response.
...
+   | Event variant | CLI action |
+   |---------------|-----------|
+   | `WindowOpened` | Print "BLE pairing window opened" to stdout |
+   | `PhoneConnected` | Print "Phone connected: {phone_id}" to stdout |
+   | `PhoneDisconnected` | Print "Phone disconnected: {phone_id}" to stdout |
+   | `PasskeyDisplay { passkey }` | Print the 6-digit zero-padded passkey to stdout and prompt `Confirm passkey? [y/N]:` on stderr |
+   | `PhoneRegistered { phone_id }` | Print "Phone registered: {phone_id}" to stdout |
+   | `WindowClosed` | Print "BLE pairing window closed" to stdout; break the event loop |
+
+3. **Passkey confirmation** — when the `PasskeyDisplay` event is received:
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in 6c73ac3.

File: docs/admin-design.md [hunk 2]
```diff
@@ -246,6 +246,46 @@ hash fields returned by the node-oriented RPC response.
...
+4. **Loop exit** — the event loop terminates when `WindowClosed` is received
+   or the stream ends. The CLI exits with code 0 on normal completion.
+
+This subcommand is exempt from `--format json` because the interactive
+passkey confirmation requires TTY access. If `--format json` is passed,
+clap rejects it (the `pairing start` subcommand does not include the format
+flag in its argument definition). If stdin is not a TTY when a passkey event
+arrives, the CLI sends `confirmed: false` to `ConfirmBlePairing` and prints
+a warning to stderr.
+
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in 6c73ac3.

File: docs/admin-design.md [hunk 3]
```diff
@@ -246,6 +246,46 @@ hash fields returned by the node-oriented RPC response.
...
+     response (`confirmed: true` or `confirmed: false`).
+4. **Loop exit** — the event loop terminates when `WindowClosed` is received
+   or the stream ends. The CLI exits with code 0 on normal completion.
+
+This subcommand is exempt from `--format json` because the interactive
+passkey confirmation requires TTY access. If `--format json` is passed,
+clap rejects it (the `pairing start` subcommand does not include the format
+flag in its argument definition). If stdin is not a TTY when a passkey event
+arrives, the CLI sends `confirmed: false` to `ConfirmBlePairing` and prints
+a warning to stderr.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 4]
```diff
@@ -68,13 +68,17 @@ to avoid collisions when tests run in parallel.
...
+**Category:** New automated (CLI process test)
 
 **Procedure:**
-1. Run `sonde-admin --help`.
-2. Assert: output contains all top-level subcommands (`node`, `program`, `schedule`, `reboot`, `ephemeral`, `status`, `state`, `modem`, `pairing`, `handler`).
-3. Run `sonde-admin node --help`.
-4. Assert: output contains nested subcommands (`list`, `get`, `register`, `remove`, `factory-reset`).
+1. Invoke `sonde-admin --help`.
+2. Assert: exit code is 0.
+3. Assert: stdout contains all top-level subcommands (`node`, `program`, `schedule`, `reboot`, `ephemeral`, `status`, `state`, `modem`, `pairing`, `handler`).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 5]
```diff
@@ -607,6 +951,90 @@ peer simulator. Structural verification that the passkey prompt calls
...
+   `OpenBlePairingRequest`.
+2. **AC-2 (passkey format):** Assert: the `PasskeyDisplay` event handler
+   formats the passkey with `{:06}` (6-digit, zero-padded).
+3. **AC-3 (prompt on stderr):** Assert: the passkey confirmation prompt
+   `Confirm passkey? [y/N]:` is written to stderr, not stdout.
+4. **AC-4 (ConfirmBlePairing):** Assert: after reading the user's `y`/`n`
+   response, the CLI calls `ConfirmBlePairing` with `confirmed: true` or
+   `confirmed: false`.
+5. **AC-5 (event printing):** Assert: each `BlePairingEvent` variant
+   (`WindowOpened`, `PhoneConnected`, `PhoneDisconnected`, `PasskeyDisplay`,
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 6]
```diff
@@ -607,6 +951,90 @@ peer simulator. Structural verification that the passkey prompt calls
...
+### T-0700a  List phones — CLI output format
+
+**Validates:** ADMIN-0702
+**Category:** New automated (CLI process test)
+
+**Procedure:**
+1. Start an admin server with no registered phones.
+2. Invoke `sonde-admin --socket <endpoint> pairing list-phones`.
+3. Assert: exit code is 0.
+4. Assert: stdout indicates no phones are registered.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 7]
```diff
@@ -282,11 +347,50 @@ to avoid collisions when tests run in parallel.
...
 **Validates:** ADMIN-0202
-**Category:** New automated
+**Category:** New automated (CLI process test)
+
+**Procedure:**
+1. Start an admin server.
+2. Invoke `sonde-admin --socket <endpoint> node register test-node 0x1234 aabbccdd` (16 hex chars, 8 bytes).
+3. Assert: exit code is non-zero.
+4. Assert: stderr contains "32 bytes" or a clear error about PSK length.
+5. Invoke `sonde-admin --socket <endpoint> node register test-node 0x1234 ZZZZ` (invalid hex).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 8]
```diff
@@ -484,6 +730,61 @@ to avoid collisions when tests run in parallel.
...
+2. Invoke `sonde-admin --socket <endpoint> state export <file1> --passphrase cli-pass`.
+3. Assert: exit code is 0 (CLI argument accepted).
+4. Set `SONDE_PASSPHRASE=env-pass` and invoke
+   `sonde-admin --socket <endpoint> state export <file2>` without `--passphrase`.
+5. Assert: exit code is 0 (env var is used as fallback).
+6. Set `SONDE_PASSPHRASE=env-pass` and invoke
+   `sonde-admin --socket <endpoint> state export <file3> --passphrase cli-pass`.
+7. Assert: exit code is 0.
+8. Import `<file3>` with `--passphrase cli-pass`.
+9. Assert: import succeeds (proving CLI arg took priority over env var).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-validation.md [hunk 9]
```diff
@@ -634,3 +1062,49 @@ peer simulator. Structural verification that the passkey prompt calls
...
+
+**Validates:** ADMIN-0800
+**Category:** New automated (CLI process test)
+
+**Procedure:**
+1. Start an admin server.
+2. Invoke `sonde-admin --socket <endpoint> handler add "*" echo hello`.
+3. Assert: exit code is 0.
+4. Assert: stdout contains "Added handler for program" and `*`.
+5. Invoke `sonde-admin --socket <endpoint> --format json handler add "aabbccdd..." echo world`.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 6c73ac3.

File: docs/admin-design.md [hunk 10]
```diff
@@ -246,6 +246,46 @@ hash fields returned by the node-oriented RPC response.
...
+     response (`accept: true` or `accept: false`).
+4. **Loop exit** — the event loop terminates when `WindowClosed` is received
+   or the stream ends. The CLI exits with code 0 on normal completion.
+
+This subcommand is exempt from `--format json` because the interactive
+passkey confirmation requires TTY access. Although `--format` is a global
+flag accepted by clap, the `pairing start` handler ignores it and always
+produces text output. If stdin is not a TTY when a passkey event
+arrives, the CLI sends `accept: false` to `ConfirmBlePairing` and prints
+a warning to stderr.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in bd268aa.

File: docs/admin-validation.md [hunk 11]
```diff
@@ -371,6 +511,71 @@ to avoid collisions when tests run in parallel.
...
+### T-0300b  Ingest program — CLI file-I/O error
+
+**Validates:** ADMIN-0300
+**Category:** New automated (CLI process test)
+
+**Procedure:**
+1. Start an admin server.
+2. Invoke `sonde-admin --socket <endpoint> program ingest /nonexistent/file.o --profile resident`.
+3. Assert: exit code is non-zero.
+4. Assert: stderr contains an I/O error message referencing the file path.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in bd268aa.

File: docs/admin-validation.md [hunk 12]
```diff
@@ -607,6 +951,91 @@ peer simulator. Structural verification that the passkey prompt calls
...
+
+**Procedure:**
+1. Start an admin server.
+2. Invoke `sonde-admin --socket <endpoint> pairing stop` with stdin piped
+   (not a TTY) and without `--yes`.
+3. Assert: exit code is non-zero (non-interactive refusal per ADMIN-0104).
+4. Invoke `sonde-admin --socket <endpoint> --yes pairing stop`.
+5. Assert: the command completes (exit code 0 or gRPC error are both
+   acceptable when no window is open, per T-0701).
+6. Assert: `--yes` bypassed the confirmation prompt (no prompt on stderr).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in bd268aa.

File: docs/admin-validation.md [hunk 13]
```diff
@@ -159,10 +213,17 @@ to avoid collisions when tests run in parallel.
 **Procedure:**
 1. Start an admin server with a registered node.
 2. Invoke `sonde-admin node remove <node-id>` with stdin connected to a PTY.
-3. Write `n\n` to stdin.
-4. Assert: exit code is non-zero and the node is not removed.
-5. Re-invoke with `--yes`.
-6. Assert: exit code is 0 and the node is removed.
+3. Assert: stderr contains `[y/N]:` prompt text.
+4. Write `n\n` to stdin.
+5. Assert: exit code is non-zero and the node is not removed.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md:218]: Fixed in 3b5a2e4.

File: docs/admin-validation.md [hunk 14]
```diff
@@ -172,9 +233,13 @@ to avoid collisions when tests run in parallel.
...
 **Procedure:**
-1. Invoke `sonde-admin node remove <node-id>` with stdin piped (not a TTY) and without `--yes`.
-2. Assert: exit code is non-zero.
-3. Assert: stderr contains "non-interactive" or "--yes".
+1. Start an admin server with a registered node.
+2. Invoke `sonde-admin node remove <node-id>` with stdin piped (not a TTY) and without `--yes`.
+3. Assert: exit code is non-zero.
+4. Assert: stderr contains "non-interactive" or "--yes".
+5. Assert: the node is not removed.
+6. Re-invoke `sonde-admin --yes node remove <node-id>` with stdin piped (not a TTY).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in 3b5a2e4.

File: docs/admin-design.md [hunk 15]
```diff
@@ -246,6 +246,46 @@ hash fields returned by the node-oriented RPC response.
...
+     response (`accept: true` or `accept: false`).
+4. **Loop exit** — the event loop terminates when `WindowClosed` is received
+   or the stream ends. The CLI exits with code 0 on normal completion.
+
+This subcommand is exempt from `--format json` because the interactive
+passkey confirmation requires TTY access. Although `--format` is a global
+flag accepted by clap, the `pairing start` handler ignores it and always
+produces text output. The current implementation does not detect non-TTY
+stdin; if stdin is piped or reaches EOF during a passkey prompt, the
+blocking read returns an empty line and the CLI sends `accept: false`.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in 58f4d10.

File: docs/admin-design.md [hunk 16]
```diff
@@ -246,6 +246,48 @@ hash fields returned by the node-oriented RPC response.
...
+   or the stream ends. The CLI exits with code 0 on normal completion.
+
+This subcommand is exempt from `--format json` because the interactive
+passkey confirmation requires TTY access. Although `--format` is a global
+flag accepted by clap, the `pairing start` handler ignores it and always
+produces text output. The current implementation does not detect non-TTY
+stdin; `read_line()` will consume piped input if available (e.g.,
+`echo y | sonde-admin …` accepts the pairing). If stdin is at EOF or the
+pipe provides no data, the read returns an empty line and the CLI sends
+`accept: false`.
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-design.md]: Fixed in e476428.

File: docs/admin-validation.md [hunk 17]
```diff
@@ -484,6 +730,61 @@ to avoid collisions when tests run in parallel.
...
+5. Assert: exit code is 0 (env var is used as fallback).
+6. Set `SONDE_PASSPHRASE=env-pass` and invoke
+   `sonde-admin --socket <endpoint> state export <file3> --passphrase cli-pass`.
+7. Assert: exit code is 0.
+8. Invoke `sonde-admin --socket <endpoint> --yes state import <file3> --passphrase cli-pass`.
+9. Assert: import succeeds (exit code 0), proving CLI arg took priority over env var.
+10. Unset `SONDE_PASSPHRASE` and invoke with stdin piped (not a TTY) and
+    without `--passphrase`.
+11. Assert: exit code is non-zero.
+12. Assert: stderr indicates that a passphrase is required (no TTY available).
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply docs/admin-validation.md]: Fixed in e476428.
